### PR TITLE
Generate prop documentation

### PR DESCRIPTION
**Pull request by colebemis**
## Problem

We're currently maintaining prop documentation manually. This leads to outdated or missing prop documentation which is frustrating for consumers. 

## Solution

This PR introduces a `Props` component that we can use in our markdown documentation files to display prop documentation.

#### `.mdx` file

```md
import {Props} from '../src/props'
import {ActionList} from '@primer/components'

## Props

<Props of={ActionList.Item} />
```

#### Output

CleanShot 2021-06-25 at 10 32 21@2x

## Next steps
- [ ] Make the `Props` component an [MDX shortcode](https://mdxjs.com/blog/shortcodes#shortcodes) so we don't have to import it at the top of every file
- [ ] Add support for Markdown syntax in prop descriptions
- [ ] Incrementally migrate component documentation to use the `Props` component

### CODE REVIEW

File: docs/gatsby-node.js [hunk 1]
```diff
@@ -35,3 +37,65 @@ exports.onCreateWebpackConfig = ({actions, plugins, loaders, getConfig}) => {
 
   actions.replaceWebpackConfig(config)
 }
+
+exports.sourceNodes = ({actions, createNodeId, createContentDigest}) => {
```
Comment (smockle) [inline docs/gatsby-node.js:41]: ✍️ Gatsby’s `sourceNodes` docs: https://www.gatsbyjs.com/docs/reference/config-files/gatsby-node/#sourceNodes

File: docs/package.json [hunk 2]
```diff
@@ -17,16 +17,18 @@
...
     "gatsby-plugin-typescript": "2.10.0",
     "prop-types": "^15.7.2",
     "react": "^17.0.0",
+    "react-docgen-typescript": "^2.0.0",
     "react-dom": "^17.0.0",
     "styled-system": "^5.1.0"
   },
   "devDependencies": {
     "cross-env": "7.0.2",
+    "globby": "^11.0.4",
```
Comment (smockle) [inline docs/package.json:31]: ✍️ [`globby`](https://www.npmjs.com/package/globby) is actively-developed, >40 million weekly downloads, <22 KB unpacked, 6 dependencies, has built-in TypeScript types, and I haven’t seen known vulnerabilities. 👍

File: docs/gatsby-node.js [hunk 3]
```diff
@@ -35,3 +37,65 @@ exports.onCreateWebpackConfig = ({actions, plugins, loaders, getConfig}) => {
...
+  const {createNode} = actions
+
+  // Extract compontent metadata from source files
+  const files = globby.sync(['../src/**/*.tsx'], {absolute: true})
+  const data = docgen.parse(files, {
+    savePropValueAsString: true,
+    propFilter: prop => {
```
Comment (smockle) [inline docs/gatsby-node.js:48]: ✍️ `react-docgen-typescript`’s `propFilter` docs: https://github.com/styleguidist/react-docgen-typescript#propfilter

File: docs/src/props.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import Paragraph from '@primer/gatsby-theme-doctocat/src/components/paragraph'
+import Table from '@primer/gatsby-theme-doctocat/src/components/table'
+import {graphql, useStaticQuery} from 'gatsby'
+import React from 'react'
+
+/** Render prop documentation for the given component */
+export function Props({of}) {
+  const data = useStaticQuery(graphql`
+    query {
+      allComponentMetadata {
```
Comment (smockle) [inline docs/src/props.js:13]: ✍️ I am not very proficient with GraphQL, so I don’t know what makes `allComponentMetadata` queryable, but quick searches will show other examples of this pattern, e.g. https://blog-nrrd.doi.gov/implementing-mdx/#Step-1-Using-Gatsbys-GraphQL-to-get-your-components

File: docs/gatsby-node.js [hunk 5]
```diff
@@ -35,3 +37,65 @@ exports.onCreateWebpackConfig = ({actions, plugins, loaders, getConfig}) => {
...
+        return Boolean(hasPropAdditionalDescription)
+      }
+
+      return true
+    }
+  })
+
+  const components = data.map(component => {
+    return {
+      name: component.displayName,
```
Comment (smockle) [inline docs/gatsby-node.js:63]: Does every component have a `displayName`?

File: src/Avatar.tsx [hunk 6]
```diff
@@ -4,7 +4,9 @@ import sx, {SxProp} from './sx'
 import {ComponentProps} from './utils/types'
 
 type StyledAvatarProps = {
+  /** Sets the width and height of the avatar. */
```
Comment (smockle) [inline src/Avatar.tsx:7]: Nice, thanks for adding comments here! ✨

File: docs/gatsby-node.js [hunk 7]
```diff
@@ -35,3 +37,65 @@ exports.onCreateWebpackConfig = ({actions, plugins, loaders, getConfig}) => {
...
+    const nodeContent = JSON.stringify(component)
+    const nodeMeta = {
+      id: createNodeId(component.name),
+      parent: null,
+      children: [],
+      internal: {
+        type: `ComponentMetadata`,
+        mediaType: `text/html`,
+        content: nodeContent,
+        contentDigest: createContentDigest(component)
```
Comment (smockle) [inline docs/gatsby-node.js:95]: ✍️ `createContentDigest` supports objects: https://github.com/gatsbyjs/gatsby/blob/f4c5c482ba5da626d4f7e2d592fdc4d43eb817ba/packages/gatsby-core-utils/src/create-content-digest.ts#L29-L30 👍

**Review by smockle — APPROVED** [review]
Let a few non-blocking comments (✍️ indicates comments which may help other reviewers).

Tested this locally:
```sh
$ cd docs
$ npm install
$ npm run clean && npm run build
$ npm run clean && npm run develop
# Visited http://localhost:8000/Avatar
```

Looks good to me! 🚢

File: docs/gatsby-node.js (hunk 5 above)
Comment (colebemis) [reply docs/gatsby-node.js:63]: Yes, according to the `react-docgen-typescript` types: https://github.com/styleguidist/react-docgen-typescript/blob/6c2a088a2ed86dc8f7b23abea50f13ff78825d76/src/parser.ts#L22

**Pull request merged**